src/Features/CSharp/Portable/ConvertForEachToFor/CSharpConvertForEachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,180 @@
+
+            // create new index varialbe name
+            var indexVariable = CreateUniqueName(foreachInfo.SemanticFacts, model, foreachStatement.Statement, "i", cancellationToken);
+
+            // put variable statement in body
+            var bodyStatement = GetForLoopBody(generator, foreachInfo, collectionVariable, indexVariable);
+
+            // create for statement from foreach statement
+            var forStatement = SyntaxFactory.ForStatement(
+                SyntaxFactory.VariableDeclaration(
+                    GetTypeExpression(
+                        generator, foreachInfo.Options, model.Compilation.GetSpecialType(SpecialType.System_Int32)),
+                    SyntaxFactory.SingletonSeparatedList(
+                        SyntaxFactory.VariableDeclarator(
+                            indexVariable.WithAdditionalAnnotations(RenameAnnotation.Create()),
+                            argumentList: default,
+                            SyntaxFactory.EqualsValueClause((ExpressionSyntax)generator.LiteralExpression(0))))),
+                SyntaxFactory.SeparatedList<ExpressionSyntax>(),
+                (ExpressionSyntax)generator.LessThanExpression(
+                    generator.IdentifierName(indexVariable),
+                    generator.MemberAccessExpression(collectionVariable, foreachInfo.CountName)),
+                SyntaxFactory.SingletonSeparatedList<ExpressionSyntax>(
+                    SyntaxFactory.PostfixUnaryExpression(
+                        SyntaxKind.PostIncrementExpression, SyntaxFactory.IdentifierName(indexVariable))),
+                bodyStatement);
+
+            if (!foreachInfo.RequireCollectionStatement)
+            {
+                // move comments before "foreach" keyword to "for". if collection statement is introduced,
+                // it should have attached to the new collection statement, so no need to do it here.
+                forStatement = forStatement.WithLeadingTrivia(foreachStatement.GetLeadingTrivia());
+            }
+
+            // replace close parenthese from "foreach" statement
+            forStatement = forStatement.WithCloseParenToken(foreachStatement.CloseParenToken);
+
+            editor.ReplaceNode(foreachStatement, forStatement);
+        }
+
+        private static TypeSyntax GetTypeExpression(
+            SyntaxGenerator generator, OptionSet options, ITypeSymbol explicitType)
+        {
+            // types are not apparent in foreach statements.
+            var isBuiltInTypeContext = TypeStyleHelper.IsBuiltInType(explicitType);
+            if (TypeStyleHelper.IsImplicitStylePreferred(
+                    options, isBuiltInTypeContext, isTypeApparentContext: false))
+            {
+                return SyntaxFactory.IdentifierName("var");
+            }
+            else
+            {
+                return (TypeSyntax)generator.TypeExpression(explicitType);
+            }
+        }
```

Comment:
Cool! I also need this method in my PR. Should there be a helper?

Reply:
TypeStyleHper is the helper :)

Reply:
we could extract this into some sort of extension on the generator.